WIP: Add compaction analysis blog post (#1982)
prometheus/docs: content/blog/2021-07-10-analyzing-compaction-efficiency.md

---

> title: "Analyzing Prometheus compaction efficiency"

Reviewer: ```suggestion
title: "Analyzing Prometheus Compaction Efficiency"
```
Title case might be better? 🤔

> created_at: 2021-07-10

Reviewer: deadline for review? =D

## The compaction process

Prometheus keeps all samples from the latest two hours of data in its memory. After the two hour period, samples currently held in memory are compacted into chunks and flushed onto disk. Each chunk contains samples from exactly one series and the maximum number of samples per chunk is 120. If a series contains more than 120 samples, Prometheus will create more than one chunk for that series. In other words, a single chunk always contains data from exactly one series, but the data from one series can be compacted into one or more chunks. 

An efficient compaction process would therefore create chunks with a number of samples as close as possible to the maximum count of 120. A chunk with less than 120 samples might, in absolute terms, use less space on disk than a chunk with exactly 120 samples. But in relative terms, a chunk with 120 samples will tend to have a better size per sample ratio than a chunk with less than 120 samples, leading to a reduction in the total amount of chunks as well as a reduction in the total disk space used by Prometheus.

## Visualizing the compaction efficiency

> In order to help visualize how well Prometheus is compacting samples, we extended the `promtool tsdb analyze` command to plot a distribution of how full chunks are relative to their maximum capacity. Running the analysis on a TSDB block consuming 4.5 bytes per sample yields the following output.

Author: Do we need a reference to promtool? I am not sure how well the community is familiar with it.

Reviewer: Yes, we need to - I think it's semi-known, but it would be no harm to reference it (even how to use it, how to build or that you can take it from release binaries)

Reviewer: ```suggestion
In order to help visualize how well Prometheus is compacting samples, we extended the `promtool tsdb analyze` command to plot a distribution of how full chunks are relative to their maximum capacity of 120 samples. Running the analysis on a TSDB block consuming the estimated 4.5 bytes per sample (instead of perfect 1.3 bytes) yields the following output.
```

```

> Fullness: Amount of chunks

Reviewer: https://github.com/prometheus/prometheus/pull/8940#pullrequestreview-697986227

     10%: #############
     20%: ######
     30%: 
     40%: 
     50%: #####################################
     60%: ######
     70%: 
     80%: 
     90%: 
    100%: ###################################
```

We see that more that a large amount of chunks are just 50% filled out. There is also a significant amount which is less than 10% full. 

As a comparison, running the analysis on a well compacted TSDB block yields the following output:
```
Compaction analysis:
Fullness: Amount of chunks
     10%: #
     20%: 
     30%: 
     40%: ####
     50%: #
     60%: ##
     70%: 
     80%: 
     90%: 
    100%: ##########################################################################################
```

> We immediatelly notice that most blocks are at full capacity. The disk usage of this block is just 1.3 bytes per sample.

Reviewer: ```suggestion
We immediatelly notice that most blocks have chunks at full capacity. The disk usage of this block is just 1.3 bytes per sample.
```

## Potential reasons for sub-optimal compaction

> In the majority of cases, the compaction process is efficient and produces chunks with the ideal size of 1-2 bytes per sample. The efficiency can, however, become degraded when high CPU load is present while compaction is taking place. Prometheus assumes that samples are scraped at regular intervals, but a lack of available CPU can lead to drift in the samples' timestamps. This in turn would reduce the efficience of compacting those samples into chunks.

Reviewer: > The efficiency can, however, become degraded when high CPU load is present while compaction is taking place.

Do you mean in scrape time? I don't think compaction CPU load can cause any issues on how chunks are compacted. What can sometimes cause it is that during compaction we used to NOT compact chunks together (unless they overlap). See: https://github.com/prometheus/prometheus/blob/d95b0972505f61be0b23b4dfedb8b4816392491b/storage/merge.go#L628

So I am not sure what exactly behavior you are trying to explain

Reviewer: Oh, another reasons is miconfigured compaction or compaction not processed fast enough etc

Author: I thought that a drift in the timestamps (e.g. scrapes are not regular), this would cause compaction to become less efficient. Is this an actual problem?

Reviewer: Definitely no. Compaction at the current state does not touch samples unless it's overlap situation (vertical merge that has to be enabled explicitly). 

I am not sure how timestamp drift in scrape time can change this too. I suppose irregular timestamps can reduce gorilla compression ration, but probably marginally.

Reviewer 2: Which version of prometheus created those blocks? This should not happen in Prometheus 2.22+
